executor/src/witgen/jit/interpreter.rs

```diff
@@ -0,0 +1,546 @@
+        &mut self,
+        expr: &SymbolicExpression<T, Variable>,
+    ) -> RPNExpression<T, usize> {
+        RPNExpression::from(&expr.map_variables(&mut |var| self.map_var(var)))
+    }
+}
+
+/// An expression in Reverse Polish Notation.
+pub struct RPNExpression<T: FieldElement, S> {
+    pub elems: Vec<RPNExpressionElem<T, S>>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum RPNExpressionElem<T: FieldElement, S> {
+    Concrete(T),
+    Symbol(S),
+    BinaryOperation(BinaryOperator),
+    UnaryOperation(UnaryOperator),
+    BitOperation(BitOperator, T::Integer),
+}
+
+impl<T: FieldElement, S: Clone> From<&SymbolicExpression<T, S>> for RPNExpression<T, S> {
+    fn from(expr: &SymbolicExpression<T, S>) -> Self {
+        fn from_inner<T: FieldElement, S: Clone>(
+            expr: &SymbolicExpression<T, S>,
+            elems: &mut Vec<RPNExpressionElem<T, S>>,
+        ) {
+            match expr {
+                SymbolicExpression::Concrete(n) => {
+                    elems.push(RPNExpressionElem::Concrete(*n));
+                }
+                SymbolicExpression::Symbol(s, _) => {
+                    elems.push(RPNExpressionElem::Symbol(s.clone()));
+                }
+                SymbolicExpression::BinaryOperation(lhs, op, rhs, _) => {
+                    from_inner(lhs, elems);
+                    from_inner(rhs, elems);
+                    elems.push(RPNExpressionElem::BinaryOperation(op.clone()));
+                }
+                SymbolicExpression::UnaryOperation(op, expr, _) => {
+                    from_inner(expr, elems);
+                    elems.push(RPNExpressionElem::UnaryOperation(op.clone()));
+                }
+                SymbolicExpression::BitOperation(expr, op, n, _) => {
+                    from_inner(expr, elems);
+                    elems.push(RPNExpressionElem::BitOperation(op.clone(), *n));
+                }
+            }
+        }
+        let mut elems = Vec::new();
+        from_inner(expr, &mut elems);
+        RPNExpression { elems }
+    }
+}
+
+impl<T: FieldElement> RPNExpression<T, usize> {
+    /// Evaluate the expression using the provided variables
+    fn evaluate(&self, stack: &mut Vec<T>, vars: &[T]) -> T {
+        self.elems.iter().for_each(|elem| match elem {
+            RPNExpressionElem::Concrete(v) => stack.push(*v),
+            RPNExpressionElem::Symbol(idx) => stack.push(vars[*idx]),
+            RPNExpressionElem::BinaryOperation(op) => {
+                let right = stack.pop().unwrap();
+                let left = stack.pop().unwrap();
+                let result = match op {
+                    BinaryOperator::Add => left + right,
+                    BinaryOperator::Sub => left - right,
+                    BinaryOperator::Mul => left * right,
+                    BinaryOperator::Div => left / right,
+                    BinaryOperator::IntegerDiv => {
+                        T::from(left.to_arbitrary_integer() / right.to_arbitrary_integer())
```

Comment:
I think it would be a big performance gain to implement division on FieldElement::Integer instead of bigint.

Reply:
Yes, exactly.